### ISSUE TITLE
Emulate a phazer with the mouse

### ISSUE THREAD

**Issue by ColinPitrat** [ColinPitrat/caprice32#34]
Detecting the clicked (or hovered ?) pixel and setting the light pen input high when the CRTC reaches it:
http://www.cpcwiki.eu/index.php/Trojan_Light_Phazer

**Comment by ColinPitrat** [comment]
Similar devices (would be nice for support to be generic enough to emulate any of them):
https://www.cpcwiki.eu/index.php/Amstrad_LP-1_light_pen
http://www.cpcwiki.eu/index.php/Amstrad_Magnum_Phaser

**Comment by ColinPitrat** [comment]
And another one mentioned in #214:
https://www.cpcwiki.eu/index.php/Gunstick

**Comment by Huntsman360** [comment]
amstrad cpc has this guns that i know:
Gunstick
Magnum
Trojan
Westphaser

Any possibility to add mouse / actual guns support? i have a sinden lightgun and will be a very nice nostalgia moment

Best regards

**Comment by ColinPitrat** [comment]
Right, the Westphaser:
https://www.cpcwiki.eu/index.php/Loriciel_Phaser_(West_Phaser)

I'll look into mouse support when time permits. It should work for Sinden lightgun too as it behaves as a mouse.

**Comment by Huntsman360** [comment]
thanks a lot, i have a arcade machine with retroarch so if you compile it in libretro core for windows i appreciate so much

Sorry for a lot a requests

**Comment by Huntsman360** [comment]
Sorry, any news on lightguns support? i know it's not prioritary but i hope some day

**Comment by ColinPitrat** [comment]
I gave it a try but couldn't find a clean way to do it. 

We'd need to activate the input (Joystick Up, Joystick Down, CRTC Light Pen Input depending on the model emulated) at this place in the code:
https://github.com/ColinPitrat/caprice32/blob/ac440f5ae7ad6a3ce5a4fcc4b4a0b2d55ab7152d/src/crtc.cpp#L1066

This would require identify which pixel we're drawing which, interestingly, is not that easy in the current code.
IIRC:
 - CRTC.sl_count gives the Y, but maybe with some subtleties (impacted by overscan?)
 - CRTC.char_count would give the X (with some calculation) with some approximation (8 pixel)

I'm not 100% sure how well this would work. I'd need a game, or even better a test program to validate it.

I'm not sure if we can do better. This code would need some better documentation which, at this point, means some reverse engineering.

I'll see if I find some time to give it a try in the coming weeks.

**Comment by Huntsman360** [comment]
thanks a lot.

**Comment by ColinPitrat** [comment]
That's not how emulation works :-) I have to emulate what the hardware is doing. The rectangles are drawn by the software running on the CPC.

What I need to do is:
 - When the mouse button is clicked, do as if a joystick button was pressed (or for the Magnum, answer on port FBFEh)
 - Continuously, when the pixel under the mouse is drawn, set the CRTC Light Pen Input (and reset when not)

My assumption that it's not possible to have a precision to the pixel is confirmed by the technical specifications:
https://www.cpcwiki.eu/index.php/Amstrad_Magnum_Phaser#Technical

```

**Comment by ColinPitrat** [comment]
I started to look into this and the first finding is that it requires specific games to support it.

For example, the Magnum Phazer was sold with 6 games supporting it:
https://cpcrulez.fr/GamesTest/compilation-magnum_light_phaser.htm

This can be found in ROMs where there are many versions and one is identified as the LIGHTGUN one:
https://cpcrulez.fr/GamesTest/operation_wolf.htm

When trying it on Caprice, every other frame is completely white. Not sure whether it's a bug in Caprice or a consequence of not having a Phazer "plugged".

Testing another game ("Mission" on the compilation disk) there's the same huge white rectangle in the whole area where you can shoot every other frame.

IIUC, these white areas should only display when pressing the phazer button?

**Comment by Huntsman360** [comment]
here in spain we have "gunstick" with games like:

Bestial Warrior
Cosmic Sheriff
El Equipo A
Guillermo Tell
Mike Gunner
Outlaws
Solo
Sootland
Space Smugglers
Target Plus
Trigger


and that i remember, when you press the trigger the white areas flash very quick and in order, and the gun if sees one, send the message to computer that can determine what objet is hit  by the frametime gun says it saw a white rectangle

**Comment by ColinPitrat** [comment]
So, good news is that I have some code that works for one gun (Amstrad Magnum Phazer) with one game from the compilation (Mission) when running in windowed mode.

I know this sounds promising, but it relies on some weird hardcoded values and there's still a lot of things that are not well understood...

**Comment by Huntsman360** [comment]
thanks a lot for the effort

It's more difficult than i thought

**Comment by Tovarichtch** [comment]
I'm currently testing for lightgun support on a frontend I'm helping with and I'm very interested to see where this is going ! Caprice32 with light gun support would be amazing for GX 4000 / CPC+. I wish I could help, but I have no coding knowledge.

**Comment by ColinPitrat** [comment]
Experimental support is now in for Amstrad Magnum Phaser (https://www.cpcwiki.eu/index.php/Amstrad_Magnum_Phaser).

To try it, press Shift+F7 and use the mouse. This requires games that specially support the Amstrad Magnum Phaser (and not another lightgun as these are likely working differently)

**Comment by Tovarichtch** [comment]
Great news !

I would love to try it, but latest release on Windows is from november 2021. Sadly, I don't know how to compile myself. If you can update the binaries so I can test it, that would be great !

**Comment by ColinPitrat** [comment]
Yeah Windows release was broken. There were multiple issues but it's now fixed:
https://github.com/ColinPitrat/caprice32/releases/tag/latest

**Comment by ColinPitrat** [comment]
Added the support for Trojan Light Phazer.
Pressing Shift+F7 once switches to the Amstrad Magnum Phazer. Pressing a second time switch to the Trojan Light Phazer.

I tested it with all the supported games:
 - Amstrad Magnum Phaser
   - Operation Wolf (clearly the best)
   - Bullseye
   - Missile Ground Zero
   - Robot Attack
   - Rookie
   - Solar Invasion
   - American Turbo-King
   - Billy the Kid
   - Bronx Street Cop
   - F-16 Fighting Falcon
   - Jungle Warfare
 - Trojan Light Phazer
   - The Enforcer
   - Skee Shoot

**Comment by Tovarichtch** [comment]
I've tested the same games as you listed above my comment, plus the other unsupported ones (West Phaser and Gunstick games). Good news is : Amstrad Magnum Phaser and Trojan Light Phazer games are working very well ! So I confirm your play test.

Trojan Light Phazer does trigger the fire for Gunstick/West Phaser light gun games. That's at least something to look into. *Almost* working.

Well done on adding light gun support !

**Comment by ColinPitrat** [comment]
Yeah Gunstick works differently. I don't yet understand how, it's very poorly documented.

**Comment by Tovarichtch** [comment]
You are right. Almost all the docs are in spanish with almost no details. But a several emulators have this integrated already like CPCEC, CPCEPower, CaPriCe Forever and even WiiTuka (for Wii). All I know is that the gun was connected to the serial port.

There's a lot of links in the description of [this youtube video](https://www.youtube.com/watch?v=C60TZxuAKSQ) with a full demonstration of the hardware [here](https://www.youtube.com/watch?v=YKyqjRJJz1k).

**Comment by ColinPitrat** [comment]
Yeah I was having a look at WiiTuka but I'm not sure whether their implementation is really faithful to the hardware. I'll check how the other emulators implement it.
